File: cdm/csharp/properties/SEScalarTime.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

public class TimeUnit
{
    private TimeUnit(string v) { Value = v; }
    protected string Value { get; }

    public static TimeUnit s { get { return new TimeUnit("s"); } }
    public static TimeUnit min { get { return new TimeUnit("min"); } }
    public static TimeUnit hr { get { return new TimeUnit("hr"); } }
    public static TimeUnit day { get { return new TimeUnit("day"); } }
    public static TimeUnit yr { get { return new TimeUnit("yr"); } }
}

public class SEScalarTime : SEScalar
{
    protected TimeUnit unit;

    public SEScalarTime() : base()
    {

    }

    public SEScalarTime(double value, TimeUnit unit)
    {
        SetValue(value, unit);
    }

    public new void SetValue(double value)
    {
        throw new System.NotSupportedException("You must provide a unit");
    }

    public void SetValue(double value, TimeUnit unit)
    {
        this.value = value;
        this.unit = unit;
    }

    public double GetValue(TimeUnit unit)
    {
        throw new System.NotImplementedException();
    }

    public TimeUnit GetUnit()
    {
        return unit;
    }
}

File: src/cpp/cdm/properties/SEScalarTime.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "cdm/CommonDefs.h"
#include "cdm/properties/SEScalarTime.h"

#pragma push_macro("Time")
#undef min
const TimeUnit TimeUnit::s("s");
const TimeUnit TimeUnit::min("min");
const TimeUnit TimeUnit::hr("hr");
const TimeUnit TimeUnit::day("day");
const TimeUnit TimeUnit::yr("yr");
#pragma pop_macro("Time")

bool TimeUnit::IsValidUnit(const std::string& unit)
{
  if (s.GetString().compare(unit) == 0)
    return true;
  if (min.GetString().compare(unit) == 0)
    return true;
  if (hr.GetString().compare(unit) == 0)
    return true;
  if (day.GetString().compare(unit) == 0)
    return true;
  if (yr.GetString().compare(unit) == 0)
    return true;
  return false;
}

const TimeUnit& TimeUnit::GetCompoundUnit(const std::string& unit)
{
  if(s.GetString().compare(unit)==0)
    return s;
  if(min.GetString().compare(unit)==0)
    return min;
  if (hr.GetString().compare(unit) == 0)
    return hr;
  if (day.GetString().compare(unit) == 0)
    return day;
  if(yr.GetString().compare(unit)==0)
    return yr;
  throw CommonDataModelException(unit + " is not a valid Time unit");
}

template class SEScalarQuantity<TimeUnit>;
